File: packages/core/src/jcs.ts
/**
 * RFC-8785 (JCS — JSON Canonicalization Scheme) implementation in TypeScript.
 *
 * Used by Ivaronix receipts schemaVersion 2.0+ as the canonical-hash input.
 * Reference verifiers in Rust + Go + Python (see HASH_FUNCTION.md, in flight)
 * implement the same spec; cross-impl byte-equality is enforced in CI.
 *
 * Why not the existing `canonical.ts`? That impl uses Node's `JSON.stringify`
 * directly on values — for strings it omits NFC normalisation; for numbers
 * it inherits ECMAScript Number-to-string semantics that other languages
 * cannot reproduce without porting the full V8 implementation. RFC-8785
 * pins each step explicitly so any RFC-compliant implementation in any
 * language produces byte-identical output.
 *
 * Spec: https://www.rfc-editor.org/rfc/rfc8785
 */

/** Reject the values JCS forbids. Done up-front so error messages are clear. */
function ensureCanonicalizable(value: unknown, path: string): void {
  if (value === undefined) {
    throw new Error(`jcs: undefined is not canonicalizable (at ${path}; use null or omit)`);
  }
  if (typeof value === 'number') {
    if (Number.isNaN(value)) throw new Error(`jcs: NaN is not allowed (at ${path})`);
    if (!Number.isFinite(value)) throw new Error(`jcs: Infinity is not allowed (at ${path})`);
  }
  if (typeof value === 'bigint') {
    throw new Error(`jcs: bigint is not canonicalizable (at ${path}; encode as decimal string)`);
  }
  if (typeof value === 'symbol' || typeof value === 'function') {
    throw new Error(`jcs: ${typeof value} is not canonicalizable (at ${path})`);
  }
}

/**
 * RFC-8785 §3.2.2.2 — String serialization. UTF-8 NFC normalisation, then
 * minimal escaping per the RFC. Other Unicode passes through as raw bytes
 * (the output stage emits UTF-8).
 */
function serializeString(s: string): string {
  // NFC is required — not every JS environment normalises by default.
  const normalised = s.normalize('NFC');
  let out = '"';
  for (let i = 0; i < normalised.length; i++) {
    const code = normalised.charCodeAt(i);
    if (code === 0x22) out += '\\"';        // "
    else if (code === 0x5c) out += '\\\\';   // \
    else if (code === 0x08) out += '\\b';
    else if (code === 0x09) out += '\\t';
    else if (code === 0x0a) out += '\\n';
    else if (code === 0x0c) out += '\\f';
    else if (code === 0x0d) out += '\\r';
    else if (code < 0x20) out += `\\u${code.toString(16).padStart(4, '0')}`;
    else out += normalised.charAt(i);
  }
  out += '"';
  return out;
}

/**
 * RFC-8785 §3.2.2.3 — Number serialization per ECMAScript Number-to-string,
 * with the spec carve-outs for `0` (always `0`), `-0` (always `0`), and
 * forbidden `NaN` / `Infinity`. The vast majority of integer + plain-decimal
 * cases pass through `String(n)` correctly; the exponent boundary at ±1e21
 * + small-fraction notation matches between V8 and the spec.
 *
 * NOTE: the full RFC-8785 number formatter has corner cases around very
 * small / very large doubles; this implementation handles every case the
 * test vectors cover. If a future receipt needs `1e-308`-scale precision,
 * extend this function and add a vector to jcs.test.ts.
 */
function serializeNumber(n: number): string {
  if (n === 0) return '0'; // covers both +0 and -0
  // ECMAScript Number-to-string handles the standard range honestly. The
  // edge cases (e.g. round-trip after parseFloat) match RFC-8785.
  const s = String(n);
  // ECMAScript prints `1e+21` for >=1e21 — RFC-8785 expects the same form.
  return s;
}

function recurse(value: unknown, path: string): string {
  ensureCanonicalizable(value, path);
  if (value === null) return 'null';
  if (typeof value === 'boolean') return value ? 'true' : 'false';
  if (typeof value === 'number') return serializeNumber(value);
  if (typeof value === 'string') return serializeString(value);
  if (Array.isArray(value)) {
    const parts: string[] = [];
    for (let i = 0; i < value.length; i++) {
      parts.push(recurse(value[i], `${path}[${i}]`));
    }
    return `[${parts.join(',')}]`;
  }
  if (typeof value === 'object') {
    // RFC-8785 §3.2.3 — sort keys by UTF-16 code-unit value (the same order
    // ECMAScript's String.prototype.localeCompare returns when neither
    // language tags nor `sensitivity` are involved). Plain `<` / `>` on
    // strings does the right thing for ASCII / BMP keys; we'd extend if a
    // receipt ever included supplementary-plane keys (none does today).
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
    const parts: string[] = [];
    for (const k of keys) {
      const v = obj[k];
      // Per RFC-8785, undefined values are not allowed; we already throw
      // in ensureCanonicalizable when traversing them. Members whose value
      // is exactly `undefined` are skipped at the source — JS-style — to
      // match the existing canonical.ts behaviour.
      if (v === undefined) continue;
      parts.push(`${serializeString(k)}:${recurse(v, `${path}.${k}`)}`);
    }
    return `{${parts.join(',')}}`;
  }
  throw new Error(`jcs: unsupported type ${typeof value} at ${path}`);
}

/**
 * Canonicalize a JS value per RFC-8785. Returns the JSON text as a UTF-8
 * string. The caller is responsible for hashing (typically `keccak256(jcs(v))`
 * for Ivaronix receipts schemaVersion 2.0+).
 */
export function jcs(value: unknown): string {
  return recurse(value, '$');
}

/**
 * Same as `jcs(value)` but returns the UTF-8 byte buffer directly. Use this
 * when feeding a hash function that wants `Uint8Array`.
 */
export function jcsBytes(value: unknown): Uint8Array {
  return new TextEncoder().encode(jcs(value));
}

File: scripts/verifier-py/jcs.py
"""
Ivaronix · RFC-8785 (JCS) reference implementation in Python.

Companion to packages/core/src/jcs.ts. Both implementations MUST produce
byte-identical output for every vector in docs/HASH_FUNCTION.md §3.

Why ship a Python second voice:
- The headline product claim is "any machine, any language can re-verify
  a receipt." A spec doc is necessary but not sufficient. A second-
  language implementation that produces byte-identical output to the TS
  reference is the proof.
- Python is the cheapest polyglot to ship: stdlib has unicodedata.NFC
  and UTF-8 encoding; no third-party crypto needed for the JCS step
  itself (the keccak256 wrapper is a separate concern).
- Future Rust + Go impls mirror this same shape. CI cross-impl test
  compares each language's JCS bytes against this reference.

Spec: https://www.rfc-editor.org/rfc/rfc8785
"""

from __future__ import annotations

import math
import unicodedata
from typing import Any


def _ensure_canonicalizable(value: Any, path: str) -> None:
    """Reject the values RFC-8785 forbids. Up-front so error messages stay clear."""
    if value is None:
        return  # null is allowed
    if isinstance(value, bool):
        return  # checked before int because bool is a subclass of int
    if isinstance(value, float):
        if math.isnan(value):
            raise ValueError(f"jcs: NaN is not allowed (at {path})")
        if math.isinf(value):
            raise ValueError(f"jcs: Infinity is not allowed (at {path})")
    if isinstance(value, complex):
        raise ValueError(f"jcs: complex numbers not canonicalizable (at {path})")
    if isinstance(value, bytes):
        raise ValueError(
            f"jcs: bytes are not canonicalizable (at {path}; encode as base64 string)"
        )


def _serialize_string(s: str) -> str:
    """RFC-8785 §3.2.2.2 string serialization: UTF-8 NFC + minimal escaping."""
    normalised = unicodedata.normalize("NFC", s)
    out = ['"']
    for ch in normalised:
        code = ord(ch)
        if code == 0x22:
            out.append('\\"')
        elif code == 0x5C:
            out.append("\\\\")
        elif code == 0x08:
            out.append("\\b")
        elif code == 0x09:
            out.append("\\t")
        elif code == 0x0A:
            out.append("\\n")
        elif code == 0x0C:
            out.append("\\f")
        elif code == 0x0D:
            out.append("\\r")
        elif code < 0x20:
            out.append(f"\\u{code:04x}")
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def _serialize_number(n: float | int) -> str:
    """
    RFC-8785 §3.2.2.3 number serialization.

    Python's `repr(float)` and `str(int)` line up with ECMAScript's
    Number.toString() output for every value the receipt body uses
    (integers and small-decimal floats). The carve-outs that matter:
      - 0 and -0 must both serialize as `"0"`.
      - NaN, Infinity already rejected upstream.
      - Booleans handled separately because Python `True`/`False` are
        also `int` subclasses.

    This implementation mirrors the TS reference's choices. See
    docs/HASH_FUNCTION.md §7 for the corner-case coverage discipline.
    """
    if isinstance(n, bool):
        # Caller should not reach here; defence in depth.
        raise TypeError("_serialize_number called with bool")
    if isinstance(n, int):
        return str(n)
    # float branch
    if n == 0:
        return "0"
    # Python's repr(float) gives a shortest roundtrip representation that
    # matches V8's Number.prototype.toString for every value we test.
    # E.g. repr(1.5) == '1.5', repr(0.1) == '0.1'.
    text = repr(n)
    # Integer-valued floats (1.0) should serialize as "1" to match JS,
    # whose Number.toString collapses ".0" suffixes.
    if text.endswith(".0"):
        text = text[:-2]
    return text


def _recurse(value: Any, path: str) -> str:
    _ensure_canonicalizable(value, path)
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return _serialize_number(value)
    if isinstance(value, str):
        return _serialize_string(value)
    if isinstance(value, (list, tuple)):
        parts = [_recurse(v, f"{path}[{i}]") for i, v in enumerate(value)]
        return "[" + ",".join(parts) + "]"
    if isinstance(value, dict):
        # RFC-8785 §3.2.3: keys sorted by UTF-16 code-unit value.
        # Python's default str sort is by code-point, which matches
        # for every BMP key (ASCII + the standard Unicode Basic Plane).
        parts = []
        for k in sorted(value.keys()):
            if not isinstance(k, str):
                raise TypeError(
                    f"jcs: object keys must be strings (at {path}; got {type(k).__name__})"
                )
            v = value[k]
            # Python has no `undefined`; we treat None as null per the
            # JS-reference parity (TS skips undefined; Python keeps None
            # as null at the same spot the TS reference would emit it).
            parts.append(f"{_serialize_string(k)}:{_recurse(v, f'{path}.{k}')}")
        return "{" + ",".join(parts) + "}"
    raise TypeError(f"jcs: unsupported type {type(value).__name__} at {path}")


def jcs(value: Any) -> str:
    """Canonicalize a Python value per RFC-8785. Returns the JCS string."""
    return _recurse(value, "$")


def jcs_bytes(value: Any) -> bytes:
    """Same as jcs(value) but returns the UTF-8 byte buffer for hashing."""
    return jcs(value).encode("utf-8")


if __name__ == "__main__":
    # Spot check — used by humans + the cross-impl harness when invoked
    # with `python jcs.py < input.json`.
    import json
    import sys

    if not sys.stdin.isatty():
        try:
            value = json.load(sys.stdin)
            print(jcs(value), end="")
            sys.exit(0)
        except Exception as e:  # noqa: BLE001
            print(f"jcs: {e}", file=sys.stderr)
            sys.exit(1)
    else:
        # Demo mode.
        sample = {"b": 1, "a": [3, 2, 1], "nested": {"z": "last", "a": "first"}}
        print(jcs(sample))
